src/pyfmi/fmi3.pyx

```diff
@@ -675,6 +678,155 @@ cdef class FMUModelBase3(FMI_BASE.ModelBase):
             raise FMUException(f"The variable {pyfmi_util.decode(variablename)} could not be found.")
         return FMIL3.fmi3_import_get_variable_base_type(variable)
 
+
+    #@enable_caching TODO: Doesnt work using enums with members with C-types, lets discuss
+    def get_model_variables(self,
+        type: Optional[FMI3_Type] = None,
+        include_alias: bool = True,
+        causality: Optional[FMI3_Causality] = None,
+        variability: Optional[FMI3_Variability] = None,
+        only_start: bool = False,
+        only_fixed: bool = False,
+        filter = None,
+        _as_list: bool = False) -> Dict[str, FMI3ModelVariable]:
+        """
+        Extract the names of the variables in a model.
+
+        Parameters::
+
+            type --
+                The type of the variables as an instance of pyfmi.fmi3.FMI3_Type.
+                Default: None (i.e all).
+
+            include_alias --
+                Currently not supported (does nothing).
+                If alias should be included or not.
+                Default: True
+
+            causality --
+                The causality of the variables as an instance of pyfmi.fmi3.FMI3_Causality.
+                Default: None (i.e all).
+
+            variability --
+                The variability of the variables as an instance of pyfmi.fmi3.FMI3_Variability.
+                Default: None (i.e all).
+
+            only_start --
+                If only variables that has a start value should be returned.
+                Default: False
+
+            only_fixed --
+                If only variables that has a start value that is fixed should be returned.
+                Default: False
+
+            filter --
+                Filter the variables using a unix filename pattern
+                matching (filter="*der*"). Can also be a list of filters
+                See http://docs.python.org/2/library/fnmatch.html.
+                Default: None
+
+        Returns::
+
+            Dict with variable name as key and a pyfmi.fmi3.FMI3ModelVariable class as value.
+        """
```

Comment:
When reviewing, might be worth to also look at the FMI2 implementation, I did a lot of clean-up as well and removed variables that looked redundant to me.